make_prg/recursion_tree.py

```diff
@@ -0,0 +1,404 @@
+"""
+This module contains the classes to explicitly represent the make_prg process recursion tree
+"""
+
+from typing import List, Set, Optional, Tuple
+from loguru import logger
+from make_prg.from_msa import MSA
+from make_prg.from_msa.cluster_sequences import kmeans_cluster_seqs, ClusteringResult
+from make_prg.utils.seq_utils import (
+    SequenceExpander,
+    remove_columns_full_of_gaps_from_MSA,
+    get_consensus_from_MSA,
+    get_number_of_unique_ungapped_sequences,
+    get_number_of_unique_gapped_sequences
+)
+from make_prg.from_msa.interval_partition import IntervalPartitioner, Interval, Intervals
+from make_prg.update.denovo_variants import UpdateData
+from make_prg.update.MLPath import MLPathError
+from abc import ABC, abstractmethod
+
+SubMSAs = List[MSA]
+
+
+class RecursiveTreeNode(ABC):
+    def __init__(self, nesting_level: int, alignment: MSA,
+                 parent: Optional["RecursiveTreeNode"], prg_builder: "PrgBuilder",
+                 children_subalignments: SubMSAs):
+        """
+        Builds a new recursive tree node.
+        Notes:
+            1. To build a node you need to know in advance how it is going to cluster (horizontally or vertically).
```

Comment:
The idea of having a `NodeFactory` class is to delegate the logic of building nodes to it. If we receive the parent node as parameter, then the logic of building the node will have to be put inside the `RecursiveTreeNode` class. This would be clearer in `c++` I think, where I would put this constructor as a private method, and `NodeFactory` as friend of the class, i.e. `NodeFactory` is the only class that is able to construct nodes. This `__init__()` method should never be called by external users, only `NodeFactory.build()` should be called, and all the logic to build nodes is inside `NodeFactory.build()`, this `__init__()` function is just a trivial constructor

Reply:
In  that case, 
```Builds a new tree node, and its children, recursively, via `NodeFactory````?

Reply:
ouch, I misunderstood your comments, sorry! I agree with them, will make the changes!

Reply:
Now that I understood your comments, I implemented your original suggestion: `Builds a new tree node, and its children, recursively` in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/b9c15a1e81ef6704581c679b7b7f23a72b764b49